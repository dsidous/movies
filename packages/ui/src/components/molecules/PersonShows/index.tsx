import React from 'react';
import { Link } from 'react-router-dom';
import Rating from '@material-ui/lab/Rating';
import { Box, Typography } from '@material-ui/core';

import { MediaImage } from '@atoms/index';
import useStyles from './styles';
import { Person_Credits } from '@typesRoots/person';

interface Props {
  shows: Person_Credits;
}

const PersonShows: React.FC<Props> = ({ shows: { cast } }) => {
  if (!cast.length) {
    return null;
  }

  const sortedShow = Array.from(new Set(cast)).sort((a, b) => {
    const dataB = b.release_date || b.first_air_date;
    const dataA = a.release_date || a.first_air_date;
    return dataB ? dataB.localeCompare(dataA) : -1;
  });

  const classes = useStyles();

  return (
    <>
      {sortedShow.map((show) => {
        const {
          id,
          poster_path,
          character,
          vote_average,
          media_type,
          title,
          name,
          release_date,
          first_air_date,
        } = show;

        const releaseDate = release_date || first_air_date;

        if (!(title || name) || releaseDate === undefined) {
          return null;
        }

        return (
          <Link to={`/${media_type}/${id}`} key={id} className={classes.root}>
            <p className={classes.poster}>
              <MediaImage
                mediaType='poster'
                size={1}
                filePath={poster_path}
                name={title || name}
              />
            </p>
            <Typography
              variant='body2'
              className={classes.title}
              component='div'
            >
              <Box fontWeight={700}>{title || name}</Box>
              {character && (
                <Typography variant='body2'>{` as ${character}`}</Typography>
              )}
            </Typography>
            {releaseDate !== '' && releaseDate !== undefined && (
              <Typography variant='body2' className={classes.release}>
                {releaseDate.substr(0, 4)}
              </Typography>
            )}
            {!!vote_average && (
              <Rating
                className={classes.rating}
                name='size-small'
                value={vote_average / 2}
                size='small'
                precision={0.5}
                readOnly
              />
            )}
          </Link>
        );
      })}
    </>
  );
};

export default PersonShows;
