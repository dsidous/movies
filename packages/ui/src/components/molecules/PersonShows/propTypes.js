import { object } from 'prop-types';

export const propTypes = {
  shows: object.isRequired,
};

export const defaultProps = {};
