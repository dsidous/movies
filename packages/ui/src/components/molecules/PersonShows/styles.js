import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      background: '#fff',
      border: '1px solid #bbb',
      borderBottom: 'none',
      display: 'grid',
      gridColumnGap: theme.spacing(2),
      gridTemplateColumns: '1fr 6fr',
      gridTemplateAreas: '"poster release" "poster title" "poster rating"',
      padding: theme.spacing(2),
      alignItems: 'center',
      justifyItems: 'start',

      [theme.breakpoints.up('sm')]: {
        gridAutoRows: theme.spacing(11),
        gridTemplateColumns: '72px 4fr 70px 100px',
        gridTemplateAreas: '"poster title release rating"',
      },

      '&:nth-child(odd)': {
        background: '#f2f2f2',
      },

      '&:last-child': {
        borderBottom: '1px solid #bbb',
      },

      '&:hover': {
        background: '#bec1c2',
        color: '#333',
      },
    },

    poster: {
      gridArea: 'poster',
      margin: 0,

      '& img': {
        display: 'block',
        height: theme.spacing(9),
        margin: '0 auto',
      },
    },

    release: {
      gridArea: 'release',
      margin: 0,

      [theme.breakpoints.up('sm')]: {
        gridColumn: '3 / 4',
      },
    },

    title: {
      gridArea: 'title',
      margin: 0,
    },

    rating: {
      gridArea: 'rating',
    },
  }),
  { name: 'personShow' },
);
