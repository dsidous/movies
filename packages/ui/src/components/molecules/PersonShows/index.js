/* eslint-disable camelcase */
import React from 'react';
import { Link } from 'react-router-dom';

import { propTypes } from './propTypes';
import MediaImage from '../../atoms/MediaImage';

const PersonShows = ({ shows: { cast } }) => {
  if (cast.length > 0) {
    const sortedShow = [].concat(cast).sort((a, b) => {
      const dataB = b.release_date || b.first_air_date;
      const dataA = a.release_date || a.first_air_date;
      return dataB ? dataB.localeCompare(dataA) : -1;
    });

    return sortedShow.map((show, i) => {
      const { id, poster_path, character } = show;
      const showAttr = show.release_date
        ? { title: 'title', releaseDate: 'release_date', showType: 'movie' }
        : { title: 'name', releaseDate: 'first_air_date', showType: 'tv' };

      const title = show[showAttr.title];
      const releaseDate = show[showAttr.releaseDate] || '';
      const { showType } = showAttr;
      const key = `${title}${id}-${i}`;
      return (
        <Link to={`/${showType}/${id}`} key={key} className="person-movie">
          <p className="person-movie__poster">
            <MediaImage
              mediaType="poster"
              size={1}
              filePath={poster_path}
              name={title}
            />
          </p>
          <p className="person-movie__release">
            {releaseDate !== '' && releaseDate !== undefined
              ? releaseDate.substr(0, 4)
              : ''}
          </p>
          <p className="person-movie__title">
            {title}
            {character && (
              <span className="person-movie__character">
                {` as ${character}`}
              </span>
            )}
          </p>
        </Link>
      );
    });
  }

  return null;
};

PersonShows.propTypes = propTypes;

export default PersonShows;
