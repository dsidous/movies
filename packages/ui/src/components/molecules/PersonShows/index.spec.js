import React from 'react';
import { shallow } from 'enzyme';
import PersonShows from '.';

const mockProps = {
  shows: {
    cast: [
      {
        id: 1,
        title: 'title',
        character: 'character',
        poster_path: 'path/to/image',
        release_date: '01.01.2010',
      },
      {
        id: 2,
        name: 'name',
        character: 'character',
        poster_path: 'path/to/image',
        first_air_date: '01.01.2011',
      },
    ],
  },
};

describe('Molecules/PersonShows', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<PersonShows {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });
});
