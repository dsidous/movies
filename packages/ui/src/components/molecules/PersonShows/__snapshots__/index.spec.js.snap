// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/PersonShows should render as expected 1`] = `ShallowWrapper {}`;
